airflow/decorators/base.py

```diff
@@ -300,6 +300,10 @@ def _infer_multiple_outputs(self):
         except TypeError:  # Can't evaluate return type.
             return False
         ttype = getattr(return_type, "__origin__", return_type)
+        if ttype == dict or ttype == Dict:
+            raise AttributeError(
+                "multiple_outputs was not set and will not implicitly unroll dict for return values"
+            )
```

Comment:
AttributeError feels wrong; maybe something like RuntimeError or a custom exception class is more suitable. I’d also want the message to be clearer (tell the user “set multiple_outputs=True is to retain the old behaviour”).

Reply:
I wonder if the new DagWarning mechanism is even better for this. It would still allow the DAG to parse (i.e. not breaking people’s existing setups) but instruct the user to fix things.

Reply:
I think it is an AttributeError. An attribute `multiple_outputs` is an attribute and was not set so we don't what to do. An AirflowException is very generic.

If we are going for option 2 I don't think the DAG should parse and in that case I think the error message should not refer to the old behavior as that implies knowing what that is. What you are describing seems more like a deprecation warning though and thus option 3?